### PR TITLE
[FEATURE] INT8 Quantization for BERT Sentence Classification and Question Answering

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
Quantization solution for BERT SC and QA with Intel DLBoost.

### Main Code Changes:
* [x] change inputs order in BERT SC dataloader to make it align with the inputs order in symbolic model(`data0=input_ids, data1=segment_ids, data2=valid_length`)
* [x] implement `BertLayerCollector` to support output clipping while calibration. Now we clip the max_range of GeLU output to 10 and the min_range of layer_norm output to -50 by default.
* [x] add calibration pass and symbolblock inference pass in `finetune_classification.py`.
* [x] add calibration pass and symbolblock inference pass in `finetune_squad.py`.
* [x] Quantization Readme
* [x] Document
* [ ] accuracy wait to remeasure

### Dependency:
https://github.com/apache/incubator-mxnet/pull/17161
https://github.com/apache/incubator-mxnet/pull/17187
https://github.com/dmlc/gluon-nlp/pull/1091
https://github.com/dmlc/gluon-nlp/issues/1127
https://github.com/dmlc/gluon-nlp/pull/1124
...

### FP32 and INT8 Accuracy:
_will remeasure on c5 when pending PRs are ready._

| Task  | maxLength | FP32 Accuracy | INT8 Accuracy | FP32 F1 | INT8 F1 |
|-------|-----------|---------------|---------------|---------|---------|
| SQUAD | 128       | 77.32         | 76.61         | 84.84   | 84.26   |
| SQUAD | 384       | 80.86         | 80.56         | 88.31   | 88.14   |
| MRPC  | 128       | 87.75         | 87.25         | 70.50   | 70.56   |

@pengzhao-intel @TaoLv @eric-haibin-lin @szha 

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: env/docker/py3.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ dependencies:
     - flaky==3.6.1
     - flake8==3.7.9
     - mock<3
-    - https://apache-mxnet.s3-us-west-2.amazonaws.com/dist/2019-12-29/dist/mxnet_cu100-1.6.0b20191229-py2.py3-none-manylinux1_x86_64.whl
+    - https://lllausen-data.s3.amazonaws.com/mxnet_cu100-1.6.0b20191231-py2.py3-none-manylinux1_x86_64.whl
```
Comment (leezu) [inline env/docker/py3.yml]: Why this change in this PR?

**Comment by leezu** [comment]
Please merge master due to #1096

**Comment by xinyu-intel** [comment]
@leezu ok, because need new API in https://github.com/apache/incubator-mxnet/pull/17161.

**Comment by leezu** [comment]
I'm removing the release-focus label because this PR depends on MXNet 1.7 and thus can't be included for the GluonNLP 0.9 release.

We can merge it soon after the release.

Thanks @xinyu-intel for the contribution!

**Comment by xinyu-intel** [comment]
@eric-haibin-lin are trying to fix the ut.

**Comment by mli** [comment]
Job PR-1080/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/12/index.html

**Comment by leezu** [comment]
@xinyu-intel unfortunately we can't update the MXNet dependency until https://github.com/apache/incubator-mxnet/issues/17292 is fixed (or we need to disable the horovod tests run on our CI)

**Comment by xinyu-intel** [comment]
@leezu thx, i'll open a new PR for update mxnet only.

**Comment by leezu** [comment]
Thanks @xinyu-intel. Even with a separate PR, we can't merge it until apache/incubator-mxnet#17292 is fixed or disable the horovod tests. 

Also we can only update the version used inside `-master.yml` files. The other files are to ensure that we don't break stable MXNet.

**Comment by mli** [comment]
Job PR-1080/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/14/index.html

**Comment by xinyu-intel** [comment]
@leezu cannot find the latest API in https://ci.mxnet.io/blue/organizations/jenkins/GluonNLP-py3-gpu-integration/detail/PR-1080/14/pipeline which tested under stable mxnet, should I skip this test in this stage?

**Comment by mli** [comment]
Job PR-1080/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/15/index.html

**Comment by TaoLv** [comment]
@leezu Any suggestion? How can we proceed?

**Comment by leezu** [comment]
@TaoLv as mentioned above https://github.com/dmlc/gluon-nlp/pull/1080#issuecomment-574714227, we currently depend on MXNet 1.6 and as this change requires MXNet 1.7 it can't be merged. After the 0.9 release we're likely to drop MXNet 1.6 requirement and this PR can be merged.

File: src/gluonnlp/calibration/__init__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (eric-haibin-lin) [inline src/gluonnlp/calibration/__init__.py]: can we add a check for mxnet version using `nlp.utils.check_version` so that mxnet < 1.7 will throw an err msg

**Comment by mli** [comment]
Job PR-1080/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/16/index.html

**Comment by eric-haibin-lin** [comment]
We can try our best to merge this for 0.9 and mark it experimental with mxnet nightly required

**Comment by xinyu-intel** [comment]
@eric-haibin-lin How about the deadline of 0.9 code freeze? We are planning to draft a document in another PR. BTW, can you point out how to skip test this file under `py3-gpu-integration`?

**Comment by mli** [comment]
Job PR-1080/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/17/index.html

**Comment by mli** [comment]
Job PR-1080/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/20/index.html

**Comment by leezu** [comment]
@xinyu-intel GluonNLP 0.9 is currently blocked until MXNet 1.6 is released. There is no date, as it requires a vote on dev list, but you still have a little time. We can keep this PR open until the documentation PR is ready if you like.

**Comment by mli** [comment]
Job PR-1080/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/21/index.html

**Comment by mli** [comment]
Job PR-1080/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/22/index.html

**Comment by mli** [comment]
Job PR-1080/23 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/23/index.html

**Comment by mli** [comment]
Job PR-1080/25 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/25/index.html

**Comment by mli** [comment]
Job PR-1080/24 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/24/index.html

**Comment by eric-haibin-lin** [comment]
@xinyu-intel is this PR ready for review?

**Comment by xinyu-intel** [comment]
@eric-haibin-lin Sure, I'll remeasure the int8 accuracy and add them to page https://gluon-nlp.mxnet.io/model_zoo/bert/index.html

**Comment by xinyu-intel** [comment]
measured fp32 accuracy with mxnet1.6rc2 and int8 accuracy with mxnet nightly.


Dataset | SQuAD 1.1 | MRPC
-- | -- | --
Model | bert_12_768_12 | bert_12_768_12
FP32 F1 / EM | 88.58 / 81.18 | 87.01 / 90.97
INT8 F1 / EM | 88.10 / 80.32 | 87.01 / 90.88

will remeasure with @zburning params later.

**Comment by zburning** [comment]
Hi, I think it should be EM 87.01 / F1 90.97?

**Comment by mli** [comment]
Job PR-1080/26 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/26/index.html

**Comment by xinyu-intel** [comment]
@zburning Thx.

There might be accuracy variance due to different SW/HW configuration.


Dataset | Model | FP32 EM | INT8 EM | FP32 F1 | INT8 F1
-- | -- | -- | -- | -- | --
SQuAD 1.1 | bert_12_768_12 | 81.18 | 80.32 | 88.58 | 88.10
MRPC | bert_12_768_12 | 87.01 | 87.01 | 90.97 | 90.88
SST | bert_12_768_12 | 93.23 | 93.00 |   |  

cc @TaoLv @eric-haibin-lin

**Comment by mli** [comment]
Job PR-1080/27 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/27/index.html

File: docs/examples/sentence_embedding/bert.md [hunk 3]
```diff
@@ -294,6 +295,98 @@ for epoch_id in range(num_epochs):
             step_loss = 0
 ```
 
+## Quantize the model
+
+GluonNLP also delivered some int8 quantization methods to improve the performance and reduce the deployment costs for the natural language inference tasks. In real production, there are two main benefits of lower precision (INT8). First, the computation can be accelerated by the low precision instruction, like Intel Vector Neural Network Instruction (VNNI). Second, lower precision data type would save the memory bandwidth and allow for better cache locality and save the power. The new feature can get up to 4X performance speedup in the latest [AWS EC2 C5 instances](https://aws.amazon.com/blogs/aws/now-available-new-c5-instance-sizes-and-bare-metal-instances/) under the [Intel Deep Learning Boost (VNNI)](https://www.intel.ai/intel-deep-learning-boost/) enabled hardware with less than 0.5% accuracy drop.
```
Comment (TaoLv) [inline docs/examples/sentence_embedding/bert.md]: Let's avoid mix using `int8` and `INT8`. How about sticking to `INT8`.

File: scripts/bert/index.rst [hunk 4]
```diff
@@ -195,6 +195,24 @@ To get the score of the dev data, you need to download the dev dataset (`dev-v2.
...
+
++-----------+-------------------+---------+---------+---------+---------+-----+---------+
+|  Dataset  | Model             | FP32 EM | INT8 EM | FP32 F1 | INT8 F1 | Log | Command |
++===========+===================+=========+=========+=========+=========+=====+=========+
+| SQuAD 1.1 | bert_12_768_12    | 81.18   | 80.32   | 88.58   | 88.10   |     |         |
++-----------+-------------------+---------+---------+---------+---------+-----+---------+
+| MRPC      | bert_12_768_12    | 87.01   | 87.01   | 90.97   | 90.88   |     |         |
++-----------+-------------------+---------+---------+---------+---------+-----+---------+
+| SST       | bert_12_768_12    | 93.23   | 93.00   |         |         |     |         |
++-----------+-------------------+---------+---------+---------+---------+-----+---------+
```
Comment (TaoLv) [inline scripts/bert/index.rst]: `SST` -> `SST-2`?

Comment (TaoLv) [reply scripts/bert/index.rst]: Also from the tables above, SST2 and MRPC are not using `EM` as a metric, but `Accuracy`.

File: scripts/bert/index.rst [hunk 5]
```diff
@@ -195,6 +195,24 @@ To get the score of the dev data, you need to download the dev dataset (`dev-v2.
 
     $ python evaluate-v2.0.py dev-v2.0.json predictions.json
 
+BERT INT8 Quantization
+~~~~~~~~~~~~~~~~~~~~~~
+
+GluonNLP provides the following example scripts to quantize fine-tuned
+BERT models into int8 data type. Note that INT8 Quantization needs a nightly
+version of mxnet-mkl.
```
Comment (TaoLv) [inline scripts/bert/index.rst]: @eric-haibin-lin Do we have any link about how to install nightly build from s3?

File: src/gluonnlp/calibration/collector.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+This collector is designed for collect and clip the layer outputs of bert
+while calibration with Gluon NLP Toolkit.
+
+@article{devlin2018bert,
+  title={BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding},
+  author={Devlin, Jacob and Chang, Ming- \
+      Wei and Lee, Kenton and Toutanova, Kristina},
+  journal={arXiv preprint arXiv:1810.04805},
+  year={2018}
+}
```
Comment (TaoLv) [inline src/gluonnlp/calibration/collector.py]: @eric-haibin-lin @leezu do we need to have this citation here?

**Comment by TaoLv** [comment]
@eric-haibin-lin Let's focus on the feature itself in this PR so we can make sure to catch the 0.9 release as an experimental feature. We can refactor the scripts and tutorials in follow-up PRs, and will be better if we get some feedbacks from users about the quantization workflow. What do you think?

File: src/gluonnlp/calibration/collector.py (hunk 6 above)
Comment (leezu) [reply src/gluonnlp/calibration/collector.py]: I don't think so

File: scripts/bert/index.rst (hunk 5 above)
Comment (eric-haibin-lin) [reply scripts/bert/index.rst]: The build wheels are here: https://apache-mxnet.s3-us-west-2.amazonaws.com/dist/index.html

**Comment by mli** [comment]
Job PR-1080/28 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1080/28/index.html
